mod-kitpvp/src/main/java/gg/solarmc/loader/kitpvp/KitPvp.java

```diff
@@ -22,66 +22,196 @@
+        if (amount <= 0) {
+            throw new IllegalArgumentException("amount must be positive");
+        }
+        KitpvpStatisticsRecord record = getStatistics(transaction);
 
-        KitpvpStatisticsRecord record = context.fetchOne(KITPVP_STATISTICS,KITPVP_STATISTICS.USER_ID.eq(userID));
-        Objects.requireNonNull(record,"kitPVP data is null! Call the police!");
+        Integer existingValue = record.getKills();
+        Integer newValue = existingValue + amount;
 
-        return record;
+        record.setKills(newValue);
+        record.store(KITPVP_STATISTICS.KILLS);
+        this.kills = newValue;
+
+        return new StatisticResult(newValue);
     }
 
+    /**
+     * Adds deaths to the user account. Infallible.
+     * @param transaction represents the transaction
+     * @param amount represents the amount of deaths to add
+     * @return a result with the new value of deaths
+     */
+    public StatisticResult addDeaths(Transaction transaction, int amount) {
+        if (amount <= 0) {
+            throw new IllegalArgumentException("amount must be positive");
+        }
+        KitpvpStatisticsRecord record = getStatistics(transaction);
+
+        Integer existingValue = record.getDeaths();
+        Integer newValue = existingValue + amount;
+
+        record.setKills(newValue);
+        record.store(KITPVP_STATISTICS.DEATHS);
+        this.deaths = newValue;
+
+        return new StatisticResult(newValue);
+    }
+
+    /**
+     * Adds assists to the user account. Infallible.
+     * @param transaction represents the transaction
+     * @param amount represents the amount of assists to add
+     * @return a result with the new value of assists
+     */
+    public StatisticResult addAssists(Transaction transaction, int amount) {
+        if (amount <= 0) {
+            throw new IllegalArgumentException("amount must be positive");
+        }
+        KitpvpStatisticsRecord record = getStatistics(transaction);
+
+        Integer existingValue = record.getAssists();
+        Integer newValue = existingValue + amount;
+
+        record.setKills(newValue);
+        record.store(KITPVP_STATISTICS.ASSISTS);
+
+        this.assists = newValue;
+
+        return new StatisticResult(newValue);
+    }
+
+    //TODO verify if the record-counting part of this (res != 0) is accurate or needs to be flipped
+
+    /**
+     * Adds an existing kit to the player's kit list
+     * @param transaction represents the transaction
+     * @param kit represents the kit to add
+     * @return whether the action was successful or not
+     */
+    public KitResult addKit(Transaction transaction, Kit kit) {
+        int res = transaction.getProperty(DSLContext.class)
+                .insertInto(KITPVP_KITS_OWNERSHIP,KITPVP_KITS_OWNERSHIP.USER_ID,KITPVP_KITS_OWNERSHIP.KIT_ID)
+                .values(userID,kit.getId())
+                .onDuplicateKeyUpdate()
+                .set(KITPVP_KITS_OWNERSHIP.KIT_ID, kit.getId())
```

Comment:
The query you are sending to the database looks like this:
```sql
INSERT INTO kitpvp_kits_ownership (userId, kitId) VALUES (?, ?) ON DUPLICATE KEY UPDATE kitId = ?
```
However, it doesn't make sense (and is redundant) to update the kitId.

So use
```
INSERT IGNORE INTO kitpvp_kits_ownership (userId, kitId) VALUES (?, ?)
```
which is the same as jooq's onDuplicateKeyIgnore

Reply:
The issue is i'm not sure if onDuplicateKeyIgnore will still update the database. I'm very aware it is redundant, but without that update (which we are checking for) it will return 1 insert no matter if it inserts correctly or not, right?

Reply:
The kitId column is already kit.getId(). It doesn't make sense to use `ON DUPLICATE KEY UPDATE kitId = ?`.

If there is a duplicate key, it means the user already owns the kit, and that there is a row with the userId and kitId. Adding `on DUPLICATE KEY UPDATE` would update that row... to again have the same kitId.

Reply:
With onDuplicateKeyIgnore, you'll get 1 if the kit was added, or 0 if the kit was already owned because there is a row with the same userId and kitId.

Reply:
Fixed